File: src/main/java/nz/laspruca/tcplugin/util/Discord.java
package nz.laspruca.tcplugin.util;

import net.dv8tion.jda.api.*;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.*;
import net.dv8tion.jda.api.hooks.*;
import net.dv8tion.jda.api.requests.*;
import net.dv8tion.jda.api.utils.cache.*;
import org.jetbrains.annotations.*;

import javax.security.auth.login.*;
import java.util.*;
import java.util.stream.*;

import static org.qrl.tcplugin.TCPlugin.*;

public class Discord extends ListenerAdapter {
	public JDA jda;
	private boolean goBrr = false;

	public Discord() {
		logger.info("Starting Discord");
		if (config.getString("prefix") == null) {
			logger.warning("No prefix set, setting to default");
			config.set("prefix", "setme");
		}
		try {
			jda = JDABuilder.createDefault(config.getString("discordToken"))
					.disableCache(CacheFlag.MEMBER_OVERRIDES, CacheFlag.VOICE_STATE)
					.enableIntents(GatewayIntent.GUILD_MEMBERS)
					.setActivity(Activity.of(Activity.ActivityType.WATCHING, "TechnoCraft Servers"))
					.addEventListeners(this, new DiscordMessageListener())
					.build();
		} catch (LoginException ex) {
			logger.warning("Unable to initialize discord api, invalid login token");
		}
	}

	public void exitDiscord() throws InterruptedException {
		if (goBrr) {
			sendMessage("Shutting down");

			jda.shutdown();
		}
	}

	public boolean logedIn() {
		return this.goBrr;
	}

	public List<String> getMembers() throws IllegalStateException {
		if (goBrr) {
			return jda
					.getGuildsByName("technocraft", true)
					.get(0)
					.loadMembers()
					.get()
					.stream()
					.map(a -> a.getNickname() == null ? a.getUser().getName() : a.getNickname())
					.collect(Collectors.toList());
		} else {
			throw new IllegalStateException("No valid connection to discord");
		}
	}

	@Override
	public void onReady(@NotNull ReadyEvent event) {
		super.onReady(event);

		goBrr = true;

		sendMessage("Server started");
	}


	public void sendMessage(String message) {
		sendMessage(new EmbedBuilder().addField(message, "", true));
	}

	public void sendMessage(EmbedBuilder message) {
		if (goBrr) {
			Objects.requireNonNull(jda.getGuildsByName("technocraft", true)
					.get(0)
					.getTextChannelById("791082936609931264"))
					.sendMessage(message.setAuthor(config.getString("prefix")).build())
					.queue();
		}
	}
}

File: src/main/kotlin/nz/laspruca/tcplugin/discord/Discord.kt
package nz.laspruca.tcplugin.discord

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.utils.cache.CacheFlag
import javax.security.auth.login.LoginException

import net.dv8tion.jda.api.events.session.ReadyEvent
import nz.laspruca.tcplugin.Plugin
import nz.laspruca.tcplugin.Plugin.Companion.discordBot
import nz.laspruca.tcplugin.Plugin.Companion.discordChannelID
import nz.laspruca.tcplugin.Plugin.Companion.discordPrefix
import nz.laspruca.tcplugin.Plugin.Companion.discordServerID
import nz.laspruca.tcplugin.Plugin.Companion.discordToken
import org.qrl.tcplugin.TCPlugin.Companion.config
import org.qrl.tcplugin.TCPlugin.Companion.logger
import kotlin.streams.toList


class Discord : ListenerAdapter() {
    private var jda: JDA? = null
    private var goBrrr = false
    val loggedIn: Boolean
        get() = goBrrr

    init {
        if (discordBot) {
            logger?.info("Starting discord")
            if (discordPrefix.isNullOrBlank()) {
                logger?.warning("No prefix set, setting to default")
                discordPrefix = "setme"
            }
            if (discordToken.isNullOrBlank()) {
                logger?.warning("No discord token set, uh oh")
            }

            try {
                jda = JDABuilder.createDefault(discordToken)
                    .disableCache(CacheFlag.MEMBER_OVERRIDES, CacheFlag.VOICE_STATE)
                    .enableIntents(GatewayIntent.GUILD_MEMBERS, GatewayIntent.GUILD_MESSAGES, GatewayIntent.MESSAGE_CONTENT, GatewayIntent.GUILD_INVITES)
                    .addEventListeners(this, DiscordMessageListener)
                    .build()
            } catch (ex: LoginException) {
                logger?.warning("Unable to initialize discord api, invalid login token")
            }

        } else {
            logger.info("Discord bot disabled "+ config.getBoolean("discordBot") + " " + discordBot)

        }

    }

    @Throws(InterruptedException::class)
    fun exitDiscord() {
        if (goBrrr && discordBot) {
            sendMessage("Shutting down")

            jda?.shutdown()
            goBrrr = false
        }
    }

    @Throws(IllegalStateException::class)
    fun getMembers(): List<String> {
        if (goBrrr && discordBot) {
            return jda!!.getGuildById(discordServerID!!)!!
                .loadMembers()
                .get()
                .stream()
                .map { a -> if (a.nickname == null) a.user.name else a.nickname!! }
                .toList()
        } else {
            throw IllegalStateException()
        }
    }

    override fun onReady(event: ReadyEvent) {
        super.onReady(event)
        goBrrr = true
        sendMessage("Server started")
    }

    fun sendMessageToChannel(message: String, channelID: String) {
        sendMessageToChannel(EmbedBuilder().addField(message, "", true), channelID)
    }

    fun sendMessageToChannel(message: EmbedBuilder, channelID: String) {
        if (goBrrr && discordBot) {
            jda!!.getGuildById(discordServerID!!)!!
                .getTextChannelById(channelID)!!
                .sendMessageEmbeds(message.setAuthor(discordPrefix).build())
                .queue()
        }
    }

    fun sendMessage(message: String) {
        sendMessage(EmbedBuilder().addField(message, "", true))
    }

    fun sendMessage(message: EmbedBuilder) {
        if (goBrrr && discordBot) {
            jda!!.getGuildById(discordServerID!!)!!
                .getTextChannelById(discordChannelID!!)!!
                .sendMessageEmbeds(message.setAuthor(discordPrefix).build())
                .queue()
        }
    }
}
